File: rs/minsk/src/analysis/text/source_text.rs
use std::ops::Index;
use std::ops::Range;

use super::line::TextLine;

pub struct SourceText {
    text: Vec<char>,
    pub lines: Vec<TextLine>,
}

impl SourceText {
    fn new(text: Vec<char>) -> Self {
        let mut source_text = Self {
            text,
            lines: Vec::new(),
        };

        source_text.parse_lines();

        source_text
    }

    fn parse_lines(&mut self) {
        let mut line_start = 0;
        let mut position = 0;

        while position < self.text.len() {
            let line_break_width = Self::get_line_break_width(&self.text, position);
            if line_break_width > 0 {
                self.add_line(line_start, position, line_break_width);
                position += line_break_width;
                line_start = position;
            } else {
                position += 1;
            }
        }

        self.add_line(line_start, position, 0);
    }

    fn get_line_break_width(text: &[char], position: usize) -> usize {
        let c = text[position];
        let l = text.get(position + 1).copied().unwrap_or('\0');
        match c {
            '\r' if l == '\n' => 2,
            '\r' | '\n' => 1,
            _ => 0,
        }
    }

    fn add_line(&mut self, line_start: usize, position: usize, line_break_width: usize) {
        let length = position - line_start;
        let length_including_line_break = length + line_break_width;
        self.lines.push(TextLine {
            start: line_start,
            length,
            length_including_line_break,
        });
    }

    pub fn line_index(&self, position: usize) -> usize {
        let mut lower = 0;
        let mut upper = self.lines.len() - 1;

        while lower <= upper {
            let index = lower + (upper - lower) / 2;
            let start = self.lines[index].start;

            if start == position {
                return index;
            }

            if start > position {
                upper = index - 1;
            } else {
                lower = index + 1;
            }
        }

        lower - 1
    }

    pub(crate) fn get(&self, offset: usize) -> Option<&char> {
        self.text.get(offset)
    }
}

impl From<Vec<char>> for SourceText {
    fn from(text: Vec<char>) -> Self {
        Self::new(text)
    }
}

impl ToString for SourceText {
    fn to_string(&self) -> String {
        self.text.iter().collect()
    }
}

impl Index<usize> for SourceText {
    type Output = char;

    fn index(&self, index: usize) -> &Self::Output {
        &self.text[index]
    }
}

impl Index<Range<usize>> for SourceText {
    type Output = [char];

    fn index(&self, index: Range<usize>) -> &Self::Output {
        &self.text[index]
    }
}

File: c2/minsk/src/code_analysis/text/source_text.c
#include "minsk/code_analysis/text/source_text.h"

#include <unicode/uchar.h>
#include <unicode/urename.h>

typedef UChar32 codepoint_t;

static bool
is_like_lf(ULineBreak prop)
{
  return prop == U_LB_MANDATORY_BREAK || prop == U_LB_LINE_FEED ||
         prop == U_LB_NEXT_LINE;
}

static size_t
get_line_break_width(string_t text, size_t position)
{
  codepoint_t current;
  codepoint_t next;
  U8_NEXT(text.data, position, text.length, current);
  U8_NEXT(text.data, position, text.length, next);

  if (current == '\r' && next == '\n')
  {
    return 2;
  }

  ULineBreak prop = u_getIntPropertyValue(current, UCHAR_LINE_BREAK);

  if (!is_like_lf(prop))
  {
    return 0;
  }

  return U8_LENGTH(current);
}

static void
add_line(
  minsk_text_line_buf_t * lines,
  size_t start,
  size_t position,
  size_t line_break_width,
  Arena * arena
)
{
  minsk_text_line_t line = {
    .start = start,
    .length = position - start,
    .length_including_line_break = position - start + line_break_width,
  };
  BUF_PUSH_ARENA(arena, lines, line);
}

static minsk_text_line_buf_t
parse_lines(string_t text, Arena * arena)
{
  minsk_text_line_buf_t lines = BUF_INIT;
  size_t position = 0;
  size_t line_start = 0;

  while (position < text.length)
  {
    size_t line_break_width = get_line_break_width(text, position);
    if (line_break_width == 0)
    {
      position++;
    }
    else
    {
      add_line(&lines, line_start, position, line_break_width, arena);
      position += line_break_width;
      line_start = position;
    }
  }

  if (position >= line_start)
  {
    add_line(&lines, line_start, position, 0, arena);
  }

  return lines;
}

extern minsk_text_source_text_t
minsk_text_source_text_from(string_t text, Arena * arena)
{
  return (minsk_text_source_text_t){
    ._arena = arena,
    .lines = parse_lines(text, arena),
    .text = text,
  };
}

extern size_t
minsk_text_source_text_get_line_index(
  minsk_text_source_text_t text,
  size_t position
)
{
  size_t lower = 0;
  size_t upper = text.lines.len - 1;

  while (lower <= upper)
  {
    size_t index = lower + (upper - lower) / 2;
    minsk_text_line_t line = text.lines.ptr[index];
    if (line.start == position)
    {
      return index;
    }
    if (line.start > position)
    {
      upper = index - 1;
    }
    else
    {
      lower = index + 1;
    }
  }
  return lower - 1;
}
